refactor(transformer): construct new AST instead of calling `ast.clone_in()`

Dunqing opened the pull request:
We should always avoid calling `.clone_in` because it will lose scoping information and it is not an efficient way. `clone_in` will traverse ast and calling `clone_in` for its all fields recursively

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9916?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#9916** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9916?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9916?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9908** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9908?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9916).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented on crates/oxc_transformer/src/es2018/object_rest_spread.rs:
```diff
@@ -979,7 +979,9 @@ impl<'a> ObjectRestSpread<'a, '_> {
             // `let { [`a`], ... rest }`
             PropertyKey::TemplateLiteral(lit) if lit.is_no_substitution_template() => {
-                let expr = Expression::TemplateLiteral(lit.clone_in(ctx.ast.allocator));
+                let quasis = ctx.ast.vec_from_iter(lit.quasis.iter().cloned());
+                let expressions = ctx.ast.vec_from_iter(lit.expressions.drain(..));
+                let expr = ctx.ast.expression_template_literal(lit.span, quasis, expressions);
```
I think this one can be simplified.

`lit.is_no_substitution_template()` has returned `true` here.

https://github.com/oxc-project/oxc/blob/dbe61c571f0894471ce237b29c8fab77229be512/crates/oxc_ast/src/ast_impl/js.rs#L526-L528

```suggestion
                let quasis = ctx.ast.vec1(lit.quasis[0].clone());
                let expressions = ctx.ast.vec();
                let expr = ctx.ast.expression_template_literal(lit.span, quasis, expressions);
```

Dunqing replied:
Good catch! I didn't notice that check

overlookmotel replied:
Note: Assuming that `is_no_substitution_template` is inlined here (which you'd expect, as it's so small), compiler should be able to optimize out the implicit bounds check in `lit.quasis[0]`, because it knows `self.quasis.len() == 1`.

That's an assumption, not a statement of fact - the compiler is sometimes much smarter than you'd expect, and sometimes much stupider! - but, generally speaking, you can expect the compiler to spot and make optimizations like that.

overlookmotel commented:
### Merge activity

* **Mar 21, 4:36 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **Mar 21, 4:36 AM EDT**: A user added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc).

Dunqing committed:
refactor(transformer): construct new AST instead of calling `ast.clone_in()` (#9916)

We should always avoid calling `.clone_in` because it will lose scoping information and it is not an efficient way. `clone_in` will traverse ast and calling `clone_in` for its all fields recursively
